src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -571,21 +575,39 @@ private void ExecuteSyntaxTreeActions(AnalysisScope analysisScope, AnalysisState
                     continue;
                 }
 
-                foreach (var analyzer in analysisScope.Analyzers)
+                var processedAnalyzers = analysisStateOpt != null ? PooledHashSet<DiagnosticAnalyzer>.GetInstance() : null;
+                try
                 {
-                    cancellationToken.ThrowIfCancellationRequested();
-
-                    ImmutableArray<SyntaxTreeAnalyzerAction> syntaxTreeActions;
-                    if (_syntaxTreeActionsMap.TryGetValue(analyzer, out syntaxTreeActions))
```

Comment:
This changes shows an example algorithmic change in execution model. Instead of iterating over all analyzers in the given `AnalysisScope` and then fetching each analyzers actions from the dictionary, we iterate over the array of `(analyzer, actions)` and execute all relevant analyzers. For command line build, analysis scope is always the full scope and does not need any additional checks. For IDE invocation, checking if an analyzer belongs to the given analysis scope is a cheap lookup.